### ISSUE TITLE
build_db error when library has many sequences

### ISSUE THREAD

**Issue by larssnip** [DerrickWood/kraken2#377]
I have made kraken2 databases for some time (years), and never experienced this until now. I used version 2.1.1:

I build a database with the human library plus my own prokaryotic libraries. This is on a computing cluster with huge amounts of memory etc. This also works fine, until the added library has many sequences! Understand me correctly: The library may be huge in terms of total amount of sequence, no problem. But, when the NUMBER OF sequences exceeds around 50 000 the build_db crashes. Below is my latest error message, anyone seen this before?
---

Building database files (step 3)...
*** Error in `build_db': munmap_chunk(): invalid pointer: 0x0000000033c516b0 ***
======= Backtrace: =========
/usr/lib64/libc.so.6(+0x7f3e4)[0x2ab2702703e4]
build_db[0x408d00]
build_db[0x40f3c3]
build_db[0x4059d2]
build_db[0x403adc]
/usr/lib64/libc.so.6(__libc_start_main+0xf5)[0x2ab270213555]
build_db[0x403fcf]
======= Memory map: ========
00400000-00417000 r-xp 00000000 00:32 134612675                          /net/fs-1/home01/larssn/bin/kraken2-master/build_db
00616000-00617000 r--p 00016000 00:32 134612675                          /net/fs-1/home01/larssn/bin/kraken2-master/build_db
00617000-00618000 rw-p 00017000 00:32 134612675                          /net/fs-1/home01/larssn/bin/kraken2-master/build_db
01f58000-33c5e000 rw-p 00000000 00:00 0                                  [heap]
2ab26faaf000-2ab26fad1000 r-xp 00000000 fd:00 33562548                   /usr/lib64/ld-2.17.so
2ab26fad1000-2ab26fad3000 rw-p 00000000 00:00 0 
2ab26fad3000-2ab26fc5c000 r-xp 00000000 00:3c 193452610                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libstdc++.so.6.0.25
2ab26fc5c000-2ab26fc66000 r--p 00188000 00:3c 193452610                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libstdc++.so.6.0.25
2ab26fc66000-2ab26fc6a000 rw-p 00192000 00:3c 193452610                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libstdc++.so.6.0.25
2ab26fc6a000-2ab26fc6e000 rw-p 00000000 00:00 0 
2ab26fc6e000-2ab26fc9b000 r-xp 00000000 00:3c 128177597                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgomp.so.1.0.0
2ab26fc9b000-2ab26fc9c000 r--p 0002c000 00:3c 128177597                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgomp.so.1.0.0
2ab26fc9c000-2ab26fc9d000 rw-p 0002d000 00:3c 128177597                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgomp.so.1.0.0
2ab26fc9d000-2ab26fcb4000 r-xp 00000000 00:3c 193452604                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgcc_s.so.1
2ab26fcb4000-2ab26fcb5000 r--p 00016000 00:3c 193452604                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgcc_s.so.1
2ab26fcb5000-2ab26fcb6000 rw-p 00017000 00:3c 193452604                  /net/fs-1/projects01/Orion/cluster/software/GCCcore/8.3.0/lib64/libgcc_s.so.1
2ab26fcb6000-2ab26fcbd000 rw-p 00000000 00:00 0 
2ab26fcd0000-2ab26fcd1000 r--p 00021000 fd:00 33562548                   /usr/lib64/ld-2.17.so
2ab26fcd1000-2ab26fcd2000 rw-p 00022000 fd:00 33562548                   /usr/lib64/ld-2.17.so
2ab26fcd2000-2ab26fcd3000 rw-p 00000000 00:00 0 
2ab26fcd3000-2ab26fdd4000 r-xp 00000000 fd:00 33562563                   /usr/lib64/libm-2.17.so
2ab26fdd4000-2ab26ffd3000 ---p 00101000 fd:00 33562563                   /usr/lib64/libm-2.17.so
2ab26ffd3000-2ab26ffd4000 r--p 00100000 fd:00 33562563                   /usr/lib64/libm-2.17.so
2ab26ffd4000-2ab26ffd5000 rw-p 00101000 fd:00 33562563                   /usr/lib64/libm-2.17.so
2ab26ffd5000-2ab26ffec000 r-xp 00000000 fd:00 33562581                   /usr/lib64/libpthread-2.17.so
2ab26ffec000-2ab2701eb000 ---p 00017000 fd:00 33562581                   /usr/lib64/libpthread-2.17.so
2ab2701eb000-2ab2701ec000 r--p 00016000 fd:00 33562581                   /usr/lib64/libpthread-2.17.so
2ab2701ec000-2ab2701ed000 rw-p 00017000 fd:00 33562581                   /usr/lib64/libpthread-2.17.so
2ab2701ed000-2ab2701f1000 rw-p 00000000 00:00 0 
2ab2701f1000-2ab2703b4000 r-xp 00000000 fd:00 33562555                   /usr/lib64/libc-2.17.so
2ab2703b4000-2ab2705b4000 ---p 001c3000 fd:00 33562555                   /usr/lib64/libc-2.17.so
2ab2705b4000-2ab2705b8000 r--p 001c3000 fd:00 33562555                   /usr/lib64/libc-2.17.so
2ab2705b8000-2ab2705ba000 rw-p 001c7000 fd:00 33562555                   /usr/lib64/libc-2.17.so
2ab2705ba000-2ab2705bf000 rw-p 00000000 00:00 0 
2ab2705bf000-2ab2705c1000 r-xp 00000000 fd:00 33562561                   /usr/lib64/libdl-2.17.so
2ab2705c1000-2ab2707c1000 ---p 00002000 fd:00 33562561                   /usr/lib64/libdl-2.17.so
2ab2707c1000-2ab2707c2000 r--p 00002000 fd:00 33562561                   /usr/lib64/libdl-2.17.so
2ab2707c2000-2ab2707c3000 rw-p 00003000 fd:00 33562561                   /usr/lib64/libdl-2.17.so
7ffdd1010000-7ffdd103b000 rw-p 00000000 00:00 0                          [stack]
7ffdd10f8000-7ffdd10fa000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
xargs: cat: terminated by signal 13
/net/fs-1/home01/larssn/bin/kraken2-master/build_kraken2_db.sh: line 143: 12664 Done                    list_sequence_files
     12665 Exit 125                | xargs -0 cat
     12666 Aborted                 (core dumped) | build_db -k $KRAKEN2_KMER_LEN -l $KRAKEN2_MINIMIZER_LEN -S $KRAKEN2_SEED_TEMPLATE $KRAKEN2XFLAG -H hash.k2d.tmp -t taxo.k2d.tmp -o opts.k2d.tmp -n taxonomy/ -m $seqid2taxid_map_file -c $required_capacity -p $KRAKEN2_THREAD_CT $max_db_flag -B $KRAKEN2_BLOCK_SIZE -b $KRAKEN2_SUBBLOCK_SIZE -r $KRAKEN2_MIN_TAXID_BITS $fast_build_flag

**Comment by larssnip** [comment]
I have investigated this more. It is NOT the number of sequences that causes the crash. It is a memory problem.

In my collection of genomes that I want to add, I have now identified 3 genomes where I can reproduce this error all the time. Here is what happens:
If I add genome 1, 2 or 3 separately, as a single genome library, everything works fine.
If I add genome 1 and 2 or 1 and 3 together, it crashes. Their order is not important, i.e. can write genome 3 and then 1 to the fasta-file, and it still crashes.
If I add genome 2 and 3 together it works fine.
Thus, genome 1 is the culprit, but ONLY when in combination with genome 2 or 3. 
I have of course investigated these genome-sequences and I cannot find anything strange, e.g. alien characters etc. In fact, genome 2 has quite many N's, but not genome 1...

Here is the error message I get (consistently) when combing genome 1 with either 2 or 3 (the print-out of the shell variables just before the crash is my own addition to build_kraken2_db.sh, for inspection):
---
The custom libraries...
Added "/net/fs-1/HumGut/krk_db/tmp/test_12.fna" to library (debug)
Building the database...
Creating sequence ID to taxonomy ID map (step 1)...
Sequence ID to taxonomy ID map complete. [0.025s]
Estimating required capacity (step 2)...
Estimated hash table requirement: 6833004 bytes
Capacity estimation complete. [0.138s]
Building database files (step 3)...
KRAKEN_KMER_LEN = 35
KRAKEN2_MINIMIZER_LEN = 31
KRAKEN2_SEED_TEMPLATE = 1111111111111111101010101010101
KRAKEN2XFLAG = 
required_capacity = 1708251
KRAKEN2_THREAD_CT = 10
max_db_flag = 
KRAKEN2_BLOCK_SIZE = 16384
KRAKEN2_SUBBLOCK_SIZE = 1639
KRAKEN2_MIN_TAXID_BITS = 0
fast_build_flag = 
*** Error in `build_db': corrupted size vs. prev_size: 0x00000000009c3580 ***
======= Backtrace: =========

…..skipping a long backtrace here…

xargs: cat: terminated by signal 13
/net/fs-1/home01/larssn/bin/kraken2-master/build_kraken2_db.sh: line 158: 48507 Done                    list_sequence_files
     48508 Exit 125                | xargs -0 cat
     48509 Aborted                 | build_db -k $KRAKEN2_KMER_LEN -l $KRAKEN2_MINIMIZER_LEN -S $KRAKEN2_SEED_TEMPLATE $KRAKEN2XFLAG -H hash.k2d.tmp -t taxo.k2d.tmp -o opts.k2d.tmp -n taxonomy/ -m $seqid2taxid_map_file -c $required_capacity -p $KRAKEN2_THREAD_CT $max_db_flag -B $KRAKEN2_BLOCK_SIZE -b $KRAKEN2_SUBBLOCK_SIZE -r $KRAKEN2_MIN_TAXID_BITS $fast_build_flag
---

Has anyone seen this before?

**Comment by larssnip** [comment]
I solved the problem. It was an issue with the taxonomy IDs. This was very difficult to reveal.

**Comment by jessikamarquis** [comment]
Hi there. Could I ask how you fixed this issue? I am getting the exact same error.

**Comment by larssnip** [comment]
Hi there. If I remember correctly, I read the nodes.dmp into R as a table, and then made certain each and every parent_tax_id (column 2) is also listed as a tax_id (column 1). This was not the case, i.e. the "parent" of some taxa were not themselves proper taxa. This was due to some NCBI tax_id's from the GTDB database was obsolete, i.e. no longer proper tax_id's. I had to look up the new tax_id for these (few) cases, and fixed it this way.

This error resulted in a segmentation fault, very difficult to trace. It is a pity that kraken2 doesn't check for this, should not be very much extra code.
